sql/plan/insubquery.go

```diff
@@ -65,50 +66,44 @@ func (in *InSubquery) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
 	switch right := in.Right.(type) {
 	case *Subquery:
 		if leftElems > 1 {
+			// TODO: support more than one element in IN
 			return nil, expression.ErrInvalidOperandColumns.New(leftElems, 1)
 		}
 
 		typ := right.Type()
-		values, err := right.EvalMultiple(ctx, row)
+		values, err := right.HashMultiple(ctx, row)
 		if err != nil {
 			return nil, err
 		}
 
-		for _, val := range values {
-			// If there are any values in the right-hand side, and the left-hand side is nil, IN evaluates to NULL
-			if leftNull {
-				return nil, nil
-			}
-
-			if !rightNull && val == nil {
-				rightNull = true
-				continue
-			}
-
-			val, err = typ.Convert(val)
-			if err != nil {
-				return nil, err
-			}
-
-			if !rightNull && val == nil {
-				rightNull = true
+		// NULL IN (list) returns NULL. NULL IN (empty list) returns 0
+		if leftNull {
+			if values.Size() == 0 {
+				return false, nil
 			}
+			return nil, nil
+		}
 
-			cmp, err := typ.Compare(left, val)
-			if err != nil {
-				return nil, err
+		val, notFoundErr := values.Get(rowKey(xxhash.New(), sql.NewRow(left)))
+		if notFoundErr != nil {
+			if _, nilValNotFoundErr := values.Get(rowKey(xxhash.New(), sql.NewRow(nil))); nilValNotFoundErr == nil {
+				return nil, nil
 			}
+			return false, nil
+		}
 
-			if cmp == 0 {
-				return true, nil
-			}
+		val, err = typ.Convert(val)
+		if err != nil {
+			return nil, err
 		}
 
-		if rightNull {
-			return nil, nil
+		cmp, err := typ.Compare(left, val)
```

Comment:
so if we get a hash collision we'll catch it here?

Reply:
Kind of... the cache either will or won't have the correct value for that hash key. Either way, the result won't be correct if there are any collisions in the subquery result. Very low chance due to 64 bit keys, but still possible. Need to fix eventually.